crates/jws/src/lib.rs

```diff
@@ -190,16 +190,262 @@ mod tests {
+            },
+            &header,
+            &payload,
+        )
+        .unwrap();
+
+        let decoded_jws = CompactJws::decode(&compact_jws).unwrap();
+
+        assert_eq!(decoded_jws.header, header);
+        assert_eq!(decoded_jws.payload, payload);
+        assert_eq!(decoded_jws.signature.is_empty(), false);
+        assert_eq!(decoded_jws.parts.len(), 3);
+    }
+
+    #[test]
+    fn test_jws_decode_incorrect_parts_error() {
         let invalid_jws = "invalid.jws";
         let result = CompactJws::decode(invalid_jws);
         assert!(matches!(result, Err(JwsError::IncorrectPartsLength(_))));
     }
 
+    #[tokio::test]
+    async fn test_jws_verify_malformed_header_alg_error() {
+        let key_manager = LocalKeyManager::new_in_memory();
+        let bearer_did = DidJwk::create(
+            Arc::new(key_manager),
+            DidJwkCreateOptions {
+                curve: Curve::Ed25519,
+            },
+        )
+        .expect("failed to create bearer did");
+
+        let key_id = bearer_did.document.verification_method[0].id.clone();
+
+        let header = JwsHeader {
+            alg: "".to_string(),
+            kid: key_id.clone(),
+            typ: "JWT".to_string(),
+        };
+        let payload = json!({
+            "sub": "1234567890",
+            "name": "John Doe",
+            "iat": 1516239022
+        })
+        .to_string()
+        .into_bytes();
+
+        let compact_jws = CompactJws::sign(
+            &bearer_did,
+            &KeySelector::KeyId {
+                key_id: key_id.clone(),
+            },
+            &header,
+            &payload,
+        )
+        .unwrap();
+
+        let result = CompactJws::verify(&compact_jws).await;
+
+        assert!(matches!(result, Err(JwsError::MalformedHeader(_))));
+    }
+
+    #[tokio::test]
+    async fn test_jws_verify_malformed_header_kid_error() {
+        let key_manager = LocalKeyManager::new_in_memory();
+        let bearer_did = DidJwk::create(
+            Arc::new(key_manager),
+            DidJwkCreateOptions {
+                curve: Curve::Ed25519,
+            },
+        )
+        .expect("failed to create bearer did");
+
+        let key_id = bearer_did.document.verification_method[0].id.clone();
+
+        let header = JwsHeader {
+            alg: "EdDSA".to_string(),
+            kid: "".to_string(),
+            typ: "JWT".to_string(),
+        };
+        let payload = json!({
+            "sub": "1234567890",
+            "name": "John Doe",
+            "iat": 1516239022
+        })
+        .to_string()
+        .into_bytes();
+
+        let compact_jws = CompactJws::sign(
+            &bearer_did,
+            &KeySelector::KeyId {
+                key_id: key_id.clone(),
+            },
+            &header,
+            &payload,
+        )
+        .unwrap();
+
+        let result = CompactJws::verify(&compact_jws).await;
+
+        assert!(matches!(result, Err(JwsError::MalformedHeader(_))));
```

Comment:
fyi I just opened #191 earlier, it's kinda relevant here, not proposing a change just making you aware